Make HDFSLogStore consistent with an Observer NameNode

Kimahriman opened the pull request:
Resolves #767

Adds an `msync` call after a successful write in the HDFSLogStore. This is needed because after a write using the FileContext API, the FileSystem API, which is used for reading, is not guaranteed to be able to read the write, even from the same process. This causes issues when a transaction is committed and then the new version is immediately read. This msync on the FileSystem API forces the cached FileSystem object to update it's internal state version so that the next read is guaranteed to see the write.

zsxwing commented:
> Also addresses the issue of creating a new FileContext object for each method call.

I'm surprised that FileContext abandoned the cache idea existing in FileSystem...

> * Just use the FileSystem API. I believe FileSystem.rename is also atomic, if the underlying file system is HDFS at least, it's just harder to interpret the results of it. Could add checks to make sure the FileSystem created is local or DistributedFileSystem

This sounds better if FileSystem.rename is atomic. We can assume HDFSLogStore is only used for HDFS.

IIUC, if RPC calls sent from two different `DFSClient`s, HDFS doesn't guarantee strong consistency when there is an Observer NameNode. This sounds surprising. Is it only read inconsistency or write may also be consistent? For example,

1. DFSClient A talking to master namenode renames a file X to Z.
2. DFSClient B talking to observer namenode renames a file Y to Z.

Is it possible that both will succeed? Is there any document explicitly describing the issue?

Kimahriman commented:
I can look into that more to figure out if the file system rename is actually atomic. But only an active namenode can serve write requests, so renames from different clients couldn't both succeed. The only potential issue is stale reads from two different clients

Kimahriman commented:
Ok learned a lot more...

The method for calling the same atomic rename that FileContext uses is protected in the FileSystem API, but it's public in DistributedFileSystem. There's been an open issue for 7 years to make this FileSystem API public: https://issues.apache.org/jira/browse/HADOOP-11452. So we can add checks for what the file system is and cast where needed, but then it complicates dealing with local file systems for testing and making sure people don't use non-atomic file systems without realizing it.

Also, there's been an open issue about caching FileContext/AbstractFileSystem objects for nearly 12 years: https://issues.apache.org/jira/browse/HADOOP-6356. 

So now I'm kinda leaning toward keeping the FileContext usage, removing the "cache" mechanism I added, and just going back to the create a new FileContext per read and write call, and fall back to super class for reads if the context can't be created. This would at least fix the inconsistency issue and can deal with/worry about the performance issue later. Still think it makes sense to have the base path as part of the constructor, but that would require more changes.

Kimahriman commented:
Removed the caching part to just fix for the consistency issue for now. Can address that later if I feel like it causes that much of an impact.

Kimahriman commented:
Yeah it's theoretically possible for that to happen, but much less likely to be an issue IMO. With an observer namenode, you configure how often it syncs with the active essentially. I think we have ours set to 10ms or so? By the time a new read starts after a write finishes there should be enough time for the observer to get up to date. And if it doesn't task retries should be able to let things wait even more, but either way that's kinda the scenario you accept when you use an observer.

The problem with the commit process is that it executes a `listFrom` _immediately_ after renaming the latest commit file on the driver, so there's basically no sync interval small enough to guarantee that will work, and this immediately fails a query with no retrying involved (even though it actually succeeded). And not using the observer for writing to delta would hurt ETL processes that are heavily reading from HDFS.

Kimahriman commented:
Seems a little hackier than consistently using the same File API

Kimahriman commented:
Added an initial attempt at supporting passing the log store path in the log store constructor. It feels a little weird though. Other option is what I initially had tried which is create a cache of scheme/authority to FileContext object (which should only ever be one), so the FileContext object just gets reused that way.

zsxwing commented:
Sorry for the delay. The current approach you proposed still has some issues:
- Adding a new parameter to LogStore is considered a breaking change as LogStore is a developer API and we would like to maintain the backward compatibility.  
- Technically the user can still hit the same issue when Spark reads the transaction logs if the sync interval is not small enough.

As you mentioned that the sync interval is usually small, how about we just pay some cost in the write method like this?

```
def write(): Unit = {
  rename the file
  if (Observer NameNode is enabled) {
   // Sleep until all of clients are synced.
    val syncInterval = get sync interval from the config
    val wakeupTime = System.currentTimeMillis() - syncInterval
    var sleepTime = wakeupTime - System.currentTimeMillis() + (depend on how the sync is implemented, we may need to add some extra value here)
    while (sleepTime > 0) {
       sleep(sleepTime)
       sleepTime = System.currentTimeMillis() - syncInterval
     }
  }
}
```
At last, we can improve our doc to mention about this behavior and suggest people to set a reasonable value to not slow the write a lot.

Kimahriman commented:
> * Adding a new parameter to LogStore is considered a breaking change as LogStore is a developer API and we would like to maintain the backward compatibility.
I attempted to add it in a backward compatible way which is why it's a little extra convoluted. There's only one or two places in the code where the log path isn't available when creating the log store. Conceptually it seems weird to _not_ include the log path as part of creating a log store (an instance of a LogStore maps to a single log on some file system). 

> * Technically the user can still hit the same issue when Spark reads the transaction logs if the sync interval is not small enough.
If you're talking about executors reading the transaction log to calculate a new snapshot, that at least can be handled by Spark retry behavior and stuff like that, and is more of the behavior you accept/sign up for with the Observer setup. Read after write consistency is assumed from the same process (just not with this mixed use of APIs).

> As you mentioned that the sync interval is usually small, how about we just pay some cost in the write method like this?
> 
> ```
> def write(): Unit = {
>   rename the file
>   if (Observer NameNode is enabled) {
>    // Sleep until all of clients are synced.
>     val syncInterval = get sync interval from the config
>     val wakeupTime = System.currentTimeMillis() - syncInterval
>     var sleepTime = wakeupTime - System.currentTimeMillis() + (depend on how the sync is implemented, we may need to add some extra value here)
>     while (sleepTime > 0) {
>        sleep(sleepTime)
>        sleepTime = System.currentTimeMillis() - syncInterval
>      }
>   }
> }
> ```
There's two sync intervals, how often the Observer NameNode will pull edits from the journal nodes, and how out of date a client will get before forcing an msync operation. The first one is the more relevant one for that snippet of code, but it's not necessarily available/needed on clients. The second one defaults to off which works fine for read-only usage patterns. Also there's no way to easily know what the client sync interval is (we use a hack that sets it using a Java system property so that we can set it to 0 (msync on every operation) on the driver and -1 (never msync after creating the initial client) on executors.

The other option which I originally did is to create a cache of scheme/authority to filecontext (which the delegating log store already does itself), I could resurrect that method too since it's already being used elsewhere.

Kimahriman commented:
I was thinking about this a little more again, how about just adding an `msync` call at the end of `writeInternal`? This would add a tiny bit of overhead to a successful write call but wouldn't affect existing read paths. Threw together something like this:

```
    try {
      val fs = path.getFileSystem(hadoopConf)
      val msync = fs.getClass.getMethod("msync")
      msync.invoke(fs)
    } catch {
      case parent: InvocationTargetException =>
        parent.getCause match {
          case _: UnsupportedOperationException | _: NoSuchMethodError =>
          case e => throw e
        }
    }
```
obviously with a thorough comment explain exactly what it's doing and why it's doing it

zsxwing commented:
@Kimahriman that's a good idea. +1. The overhead should be tiny since this only happens in the write path which is not a cheap operation anyway.

Kimahriman commented:
Updated to that, let me know if I went overboard on the comment explaining it hah

zsxwing reviewed (changes requested):
We are trying to create a separate module for all Log Store implementations so that we can share it in the entire ecosystem. Could you also update https://github.com/delta-io/delta/blob/master/storage/src/main/java/io/delta/storage/HDFSLogStore.java ?

In addition, it would be create to create a separate method for this hack. For example,

```
/**
<put your comment here>
*/
private def msyncIfSupported(hadoopConf: Configuration) {
    try {
      val fs = path.getFileSystem(hadoopConf)
      val msync = fs.getClass.getMethod("msync")
      msync.invoke(fs)
    } catch {
      case parent: InvocationTargetException =>
        parent.getCause match {
          case _: UnsupportedOperationException | _: NoSuchMethodError =>
          case e => throw e
        }
    }
}
``` 

Could you also update the PR description to reflect the latest change?

Regarding the test, any thoughts on how to prevent us from breaking it?

Kimahriman commented:
Yeah I can update that as well and make it it's own function. I have absolutely no idea how to test it other than try mock the msync method and make sure it gets called

Kimahriman commented:
Updated it and made it it's own function, and updated the description. I wasn't sure the best way to handle the Java exception handling. Also still not really sure how to test it. I could maybe try to create another separate test file system that does some kind of fancy thing with timestamps or something to pretend it can't read things without an msync call

zsxwing commented on storage/src/main/java/io/delta/storage/HDFSLogStore.java:
```diff
@@ -137,6 +139,37 @@ private void writeInternal(
+            FileSystem fs = path.getFileSystem(hadoopConf);
+            Method msync = fs.getClass().getMethod("msync");
+            msync.invoke(fs);
+        } catch (NoSuchMethodException e) {
+        } catch (IllegalAccessException e) {
+        } catch (InvocationTargetException e) {
```
We don't need these as the one below should handle them

Kimahriman replied:
Should I just not throw in any case and add a comment?

```
} catch (Throwable e) {
  // Best effort so ignore errors
}
```
?

zsxwing commented on storage/src/main/java/io/delta/storage/HDFSLogStore.java:
```diff
@@ -137,6 +139,37 @@ private void writeInternal(
+        } catch (IllegalAccessException e) {
+        } catch (InvocationTargetException e) {
+        } catch (Throwable e) {
+            if (!LogStoreErrors.isNonFatal(e)) {
```
nit: it's better to add a comment such as `We ignore non fatal errors as calling msync is best effort.`

Kimahriman replied:
Same as above, Java doesn't seem to recognize that isNonFatal is catching all those cases and still complains that the exception must be declared to be thrown

zsxwing replied:
How about this
```
    private void msyncIfSupported(Path path, Configuration hadoopConf) throws IOException {
        try {
            FileSystem fs = path.getFileSystem(hadoopConf);
            Method msync = fs.getClass().getMethod("msync");
            msync.invoke(fs);
        } catch (InterruptedIOException e) {
            throw e;
        } catch (Throwable e) {
            if (e instanceof InterruptedException) {
                // Propagate the interrupt status
                Thread.currentThread().interrupt();
            }
            // We ignore non fatal errors as calling msync is best effort.
        }
    }
```
We don't want to ignore interrupt status as it's a signal to stop running. The rest should be okay to ignore.

Kimahriman replied:
Java seems happy so works for me. Updated scala to just ignore NonFatal

zsxwing commented on core/src/main/scala/org/apache/spark/sql/delta/storage/HDFSLogStore.scala:
```diff
@@ -124,6 +125,37 @@ class HDFSLogStore(sparkConf: SparkConf, defaultHadoopConf: Configuration)
+      val msync = fs.getClass.getMethod("msync")
+      msync.invoke(fs)
+    } catch {
+      case parent: InvocationTargetException =>
+        parent.getCause match {
+          case _: UnsupportedOperationException =>
+          case e => throw e
+        }
+      case _: NoSuchMethodException =>
```
Could you change this to match the java version?

Kimahriman replied:
Yeah I can once figuring out the right java approach

Kimahriman commented:
I attempted to add a test using a new custom file system using timestamps to act as the state ID, seems to be working

zsxwing commented on storage/src/main/java/io/delta/storage/HDFSLogStore.java:
```diff
@@ -18,6 +18,9 @@
 import java.io.FileNotFoundException;
 import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
```
nit: remove these two unused imports

Kimahriman replied:
Method is used, but I'll remove the other one

zsxwing commented on core/src/test/scala/org/apache/spark/sql/delta/TimestampLocalFileSystem.scala:
```diff
@@ -0,0 +1,71 @@
+import org.apache.hadoop.fs.FileStatus
+
+/**
+ * This custom fs implementation is used for testing the msync calling in HDFSLogStore writes
```
nit: add a simple comment to explain how it works
```
If its `msync` is not called, `listStatus` will return a stale result.
```

Kimahriman replied:
will add

zsxwing commented on core/src/test/scala/org/apache/spark/sql/delta/LogStoreSuite.scala:
```diff
@@ -301,6 +301,30 @@ trait HDFSLogStoreSuiteBase extends LogStoreSuiteBase {
+        // Initialize the TimestampLocalFileSystem object which will be reused later due to the
+        // FileSystem cache
+        assert(store.listFrom(path, sessionHadoopConf).length == 0)
+        // The LocalFileSystem only tracks modified time at second granularity, so we need to
+        // wait a second to make sure the behavior works
+        Thread.sleep(1000)
```
why do we need `sleep` here?

Kimahriman replied:
Basically (at least on my machine) the RawLocalFileSystem's FileStatus.getModificationTime is only returning second-granular times. Without it I ran into the issue of:

- FileSystem gets initialized and sets `latestTimestamp` to 53217 milliseconds
- `store.write` creates the file at 53321 milliseconds
- `listStatus` has a `getModificationTime` of that file as 53000
- 53000 < 53217 so it's not considered "stale" even though it was written after 53217

So the sleep makes sure the modification time at least gets to 54000. Obviously those numbers are made up. Tried to convey that in the comment, let me know if you can think of a better description

Kimahriman replied:
Added a little more to the comment

zsxwing commented on core/src/test/scala/org/apache/spark/sql/delta/LogStoreSuite.scala:
```diff
@@ -301,6 +301,30 @@ trait HDFSLogStoreSuiteBase extends LogStoreSuiteBase {
+        Thread.sleep(1000)
+
+        store.write(path, Iterator("zero", "none"), overwrite = false, sessionHadoopConf)
+        assert(store.listFrom(path, sessionHadoopConf).length == 1)
```
nit: add a comment above to explain how it works
```
Verify `msync` is called by checking whether `listFrom` return the latest result.
```

zsxwing reviewed (commented):
Cool! Left some style comments. Otherwise LGTM

zsxwing commented on core/src/test/scala/org/apache/spark/sql/delta/TimestampLocalFileSystem.scala:
```diff
@@ -0,0 +1,72 @@
+
+  override def initialize(name: URI, conf: Configuration): Unit = {
+    uri = URI.create(name.getScheme + ":///")
+    latestTimestamp = System.currentTimeMillis()
```
If we remove this, do we still need `Thread.sleep(1000)` in the test (in other words, initialize `latestTimestamp` to 0)? My concern is `Thread.sleep(1000)` is not guaranteed to sleep 1000ms, which may make this test flaky.

Kimahriman replied:
I added this to be slightly more representative of a file system msync'ing on startup, but if I just set this to 0 then I think the existing test _would_ still pass without the sleep so I can just do that

zsxwing reviewed (approved):
LGTM. Thanks for your patience!
